### PR TITLE
Filter for ArrayObject, SplObjectStorage collections and associative collections

### PR DESCRIPTION
**Pull request by semmel**
This is a first attempt to add `ArrayObject` and `SplObjectStorage` to the list of supported collections.

In difference to [Ramda](http://ramdajs.com/) for JavaScript I decided to provide a separate `filter_values` function for associative collections. Thus user of the library must explicitly choose the appropriate function suitable for his type of data. 

This is because in PHP [determining that a collection is associative or not](http://stackoverflow.com/questions/5996749/determine-whether-an-array-is-associative-hash-or-not) seems to be [quite expensive](https://gist.github.com/Thinkscape/1965669).  

- `filter` now works on non-associative `ArrayObject` and the object items of `SplObjectStorage`.
- `filter_values` works on associative `array`, the values in  `ArrayObject` and the data items of `SplObjectStorage`. This is in the same spirit as *Ramda's* `filter` works on JavaScript dictionary objects `{key: "value", key2: "value2"}`.

The implementation is based on a new internal reduce algorithm `_iterable_reduce()` for `Iterators`. By providing another reducer to `_iterable_reduce` `map` can be extended for other collections as well.

How do I generate the documentation markdown files?

### CODE REVIEW

**Comment by webNeat** [comment]
Hello @semmel . I am very happy that you are interested on this library.

I have written a [contribution guide](https://github.com/tarsana/functional/blob/master/CONTRIBUTING.md) for you to understand how is this library structured and how to build and generate docs and tests (yes, most of the tests are generated from the source code).

About deciding if an `array` is a simple list (Array in JS) or an associative array (Object in JS), the function [type](https://github.com/tarsana/functional/blob/master/docs/common.md#type) can be used for that (sure it's slow if the array is very large).

**Comment by semmel** [comment]
Hi @webNeat the contribution guide is very valuable, thank you!

Regarding the detection if an array is an associative array, I doubt that this can be done in PHP. In [this discussion on Stackoverflow](http://stackoverflow.com/questions/173400/how-to-check-if-php-array-is-associative-or-sequential) different criteria are discussed:
* Firstly, how to determine whether an array has only numeric keys (like `F\type` works I guess)
* Secondly, how to determine whether an array has sequential numeric keys, starting from 0 

I think both points are misleading. The user of *Functional* should know what he intends to accomplish with that particular collection and decide to use it in associative fashion or just as a list.

What about the case when I need integers as keys in my associative collection? E.g. as id's for a store of objects or a distribution of some integer parameter like age, placement, number of items purchased... 
```php
$purchases = [0 => 0.6, 1=> 0.3, 2=>0.05, 3=>0.05];
// verify that the distribution sum is 100%:
// correct results:
Tarsana\Functional\type($purchases); // => 1
Tarsana\Functional\reduce("Tarsana\Functional\plus", 0, $purchases); //=>1
Tarsana\Functional\take(1, $purchases); // => array(1) {  [0] =>  double(0.6) }
// But why?
Tarsana\Functional\type($purchases); // List (not Array!)
Tarsana\Functional\head($purchases); // => 0.6 ???
```
Most functions return the desired result because *Functional* dispatches to Php's `array_` functions.

In summary:
* Maybe that `Tarsana\Functional\type( [0 => 0.6, 1=> 0.3, 2=>0.05, 3=>0.05]); // List` is an error and there should be a separate Issue?
* Php `array_` functions "automagically" do the right thing for `array`and associative `array`. Should *Functional* support other collections as well?
